lib/attestations/apple.js

```diff
@@ -0,0 +1,108 @@
+// appleStmtFormat = {
+//    x5c: [ credCert: bytes, * (caCert: bytes) ]
+// }
+function appleParseFn(attStmt) {
+	// Step 1: Verify that attStmt is valid CBOR conforming to the syntax
+	// defined above and perform CBOR decoding on it to extract the contained fields.
+	const ret = new Map();
+
+	const x5c = attStmt.x5c;
+	if (!Array.isArray(x5c)) {
+		throw new TypeError(
+			"expected apple attestation x5c field to be of type Array"
+		);
+	}
+
+	const abX5c = [];
+	for (let cert of x5c) {
+		cert = coerceToArrayBuffer(cert, "apple x5c cert");
+		abX5c.push(cert);
+	}
+
+	ret.set("credCert", abX5c.shift());
+
+	return ret;
+}
+
+async function appleValidateFn() {
+	const parsedCredCert = this.authnrData.get("credCert");
+
+	// Step 2: Concatenate authenticatorData(rawAuthnrData) and clientDataHash(rawClientData) to form nonceToHash.
+	const rawClientData = this.clientData.get("rawClientDataJson");
+	const rawAuthnrData = this.authnrData.get("rawAuthnrData");
+
+	const clientDataHash = await tools.hashDigest(rawClientData);
+
+	const rawAuthnrDataBuf = new Uint8Array(rawAuthnrData);
+	const clientDataHashBuf = new Uint8Array(clientDataHash);
+
+	const nonceToHash = appendBuffer(rawAuthnrDataBuf, clientDataHashBuf);
+
+	// Step 3: Perform SHA-256 hash of nonceToHash to produce nonce.
+	const nonce = await tools.hashDigest(nonceToHash);
+
+	// Step 4: Verify that nonce === value of extension with key OID 1.2.840.113635.100.8.2
+	const credCert = new Certificate(
+		coerceToBase64(parsedCredCert, "parsedCredCert")
+	);
+	this.audit.journal.add("credCert");
+	const extensions = credCert.getExtensions();
+	let expectedNonce;
+	for (const ext of extensions) {
+		if (Array.isArray(ext) && ext.length > 1) {
+			if (ext[0] === "1.2.840.113635.100.8.2") {
+				if (Array.isArray(ext[1]) && ext[1].length) {
+					expectedNonce = ext[1][0];
+				}
+			}
+		}
+	}
+	if (!expectedNonce) {
+		throw new Error(
+			"extension with key '1.2.840.113635.100.8.2' (apple) was not found"
+		);
+	}
+
+	if (!arrayBufferEquals(expectedNonce, nonce)) {
+		throw new Error("nonce did not match expectedNonce");
+	}
+
+	// Step 5: Verify that the credential public key equals the Subject Public Key of credCert.
+	const credentialPublicKey = new PublicKey();
+	await credentialPublicKey.fromPem(
+		this.authnrData.get("credentialPublicKeyPem")
+	);
+	console.log(await credCert.getPublicKey());
+	const certificatePublicKey = new PublicKey();
+	certificatePublicKey.fromCryptoKey(await credCert.getPublicKey());
+	const credentialPublicKeyReexportedPem = await credentialPublicKey.toPem(
+		true
+	);
+	const certificatePublicKeyReexportedPem = await certificatePublicKey.toPem(
+		true
+	);
+	this.audit.journal.add("credentialPublicKeyPem");
+
+	if (credentialPublicKeyReexportedPem !== certificatePublicKeyReexportedPem) {
+		throw new Error("certificatePublicKey did not match credentialPublicKey");
+	}
+
+	// ToDo: Verify certificate chain?
```

Comment:
Can you elaborate why this is not needed for safe validation in the beginning? People probably assume that we do all cryptographic validations if we say that we are compatible.

Reply:
Wierd thing is that the spec says nothing about this

https://www.w3.org/TR/webauthn-2/#sctn-apple-anonymous-attestation

It almost seems like the RP is supposed to do this?

Reply:
Good point - I dont see a specific validation step for this either, so I'd say we're good by following the standard. Can we remove the comment then to remove confusion?

Reply:
👍 

Submitted a "third party" PR

https://github.com/randomprogramming/fido2-lib/pull/4/files